// ngx-responsive/esm2015/@directives/useragent/useragent-info.directive.js
import * as tslib_1 from "tslib";
/**
 * @name useragent-info.directive
 * @description Useragent info directive in ngx-responsive
 *
 * @license MIT
 */
import { Output, EventEmitter, Directive, ChangeDetectorRef } from '@angular/core';
import { PLATFORM_ID, Inject } from '@angular/core';
import { ResponsiveState } from '../../@core/providers/responsive-state/responsive-state';
import { UserAgentInfo } from './useragent-info';
let UserAgentInfoDirective = class UserAgentInfoDirective extends UserAgentInfo {
    constructor(_responsiveState, cd, _platformId) {
        super(_responsiveState, _platformId);
        this._responsiveState = _responsiveState;
        this.cd = cd;
        this._platformId = _platformId;
        this.info = new EventEmitter();
    }
    ngOnInit() {
        this.connect();
    }
    ngOnDestroy() {
        this.disconnect();
    }
    _emitUserAgent(value) {
        this.info.emit(value);
        this.cd.markForCheck();
    }
};
tslib_1.__decorate([
    Output(),
    tslib_1.__metadata("design:type", EventEmitter)
], UserAgentInfoDirective.prototype, "info", void 0);
UserAgentInfoDirective = tslib_1.__decorate([
    Directive({
        selector: 'user-agent-info'
    }),
    tslib_1.__param(2, Inject(PLATFORM_ID)),
    tslib_1.__metadata("design:paramtypes", [ResponsiveState,
        ChangeDetectorRef, Object])
], UserAgentInfoDirective);
export { UserAgentInfoDirective };
