// ngx-responsive/esm5/@directives/responsive-window/responsive-window.js
import * as tslib_1 from "tslib";
/**
 * @name responsive-window
 * @description responsiveWindow Directive in ngx-responsive
 *
 * @license MIT
 */
import { Directive, Input, ElementRef, ChangeDetectorRef } from '@angular/core';
import { PLATFORM_ID, Inject } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { ResponsiveState } from '../../@core/providers/responsive-state/responsive-state';
var ResponsiveWindowDirective = /** @class */ (function () {
    function ResponsiveWindowDirective(_responsiveState, el, cd, _platformId) {
        this._responsiveState = _responsiveState;
        this.el = el;
        this.cd = cd;
        this._platformId = _platformId;
        this._isBrowser = null;
        this._isBrowser = isPlatformBrowser(this._platformId);
        if (this._isBrowser) {
            this.element = el.nativeElement;
        }
    }
    ResponsiveWindowDirective.prototype.ngOnInit = function () {
        if (this._isBrowser) {
            this._responsiveState.registerWindow(this);
        }
    };
    ResponsiveWindowDirective.prototype.ngDoCheck = function () {
        if (this._isBrowser) {
            var _update = this._ifValueChanged(this._noRepeat, this.name);
            if (_update) {
                this._responsiveState.unregisterWindow(this);
                this._responsiveState.registerWindow(this);
                this.cd.markForCheck();
            }
        }
    };
    ResponsiveWindowDirective.prototype.ngOnDestroy = function () {
        if (this._isBrowser) {
            this._responsiveState.unregisterWindow(this);
        }
    };
    ResponsiveWindowDirective.prototype.getWidth = function () {
        return (this._isBrowser) ? this.element.offsetWidth : 0;
    };
    ResponsiveWindowDirective.prototype._ifValueChanged = function (oldValue, newValue) {
        if (oldValue === newValue) {
            return false;
        }
        else {
            this._noRepeat = newValue;
            return true;
        }
    };
    tslib_1.__decorate([
        Input('responsive-window'),
        tslib_1.__metadata("design:type", String)
    ], ResponsiveWindowDirective.prototype, "name", void 0);
    ResponsiveWindowDirective = tslib_1.__decorate([
        Directive({
            selector: "[responsive-window]",
            exportAs: "container"
        }),
        tslib_1.__param(3, Inject(PLATFORM_ID)),
        tslib_1.__metadata("design:paramtypes", [ResponsiveState,
            ElementRef,
            ChangeDetectorRef, Object])
    ], ResponsiveWindowDirective);
    return ResponsiveWindowDirective;
}());
export { ResponsiveWindowDirective };
