// ngx-responsive/esm2015/@directives/responsive-window/responsive-window.js
import * as tslib_1 from "tslib";
/**
 * @name responsive-window
 * @description responsiveWindow Directive in ngx-responsive
 *
 * @license MIT
 */
import { Directive, Input, ElementRef, ChangeDetectorRef } from '@angular/core';
import { PLATFORM_ID, Inject } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { ResponsiveState } from '../../@core/providers/responsive-state/responsive-state';
let ResponsiveWindowDirective = class ResponsiveWindowDirective {
    constructor(_responsiveState, el, cd, _platformId) {
        this._responsiveState = _responsiveState;
        this.el = el;
        this.cd = cd;
        this._platformId = _platformId;
        this._isBrowser = null;
        this._isBrowser = isPlatformBrowser(this._platformId);
        if (this._isBrowser) {
            this.element = el.nativeElement;
        }
    }
    ngOnInit() {
        if (this._isBrowser) {
            this._responsiveState.registerWindow(this);
        }
    }
    ngDoCheck() {
        if (this._isBrowser) {
            const _update = this._ifValueChanged(this._noRepeat, this.name);
            if (_update) {
                this._responsiveState.unregisterWindow(this);
                this._responsiveState.registerWindow(this);
                this.cd.markForCheck();
            }
        }
    }
    ngOnDestroy() {
        if (this._isBrowser) {
            this._responsiveState.unregisterWindow(this);
        }
    }
    getWidth() {
        return (this._isBrowser) ? this.element.offsetWidth : 0;
    }
    _ifValueChanged(oldValue, newValue) {
        if (oldValue === newValue) {
            return false;
        }
        else {
            this._noRepeat = newValue;
            return true;
        }
    }
};
tslib_1.__decorate([
    Input('responsive-window'),
    tslib_1.__metadata("design:type", String)
], ResponsiveWindowDirective.prototype, "name", void 0);
ResponsiveWindowDirective = tslib_1.__decorate([
    Directive({
        selector: "[responsive-window]",
        exportAs: "container"
    }),
    tslib_1.__param(3, Inject(PLATFORM_ID)),
    tslib_1.__metadata("design:paramtypes", [ResponsiveState,
        ElementRef,
        ChangeDetectorRef, Object])
], ResponsiveWindowDirective);
export { ResponsiveWindowDirective };
